### PR TITLE
Various improvements to AI-related logic (Canopy lands, Ugin's Labyrinth, Zuran Orb, Pyrite Spellbomb, Eladamri)

### PR DESCRIPTION
**Pull request by Agetian**
Some work towards #5604.

### CODE REVIEW

File: forge-ai/src/main/java/forge/ai/ability/ChooseCardAi.java [hunk 1]
```diff
@@ -58,11 +59,27 @@ protected boolean checkApiLogic(final Player ai, final SpellAbility sa) {
     protected boolean checkAiLogic(final Player ai, final SpellAbility sa, final String aiLogic) {
         final Card host = sa.getHostCard();
         final Game game = ai.getGame();
+
         ZoneType choiceZone = ZoneType.Battlefield;
+        CardCollectionView choices = new CardCollection();
         if (sa.hasParam("ChoiceZone")) {
-            choiceZone = ZoneType.smartValueOf(sa.getParam("ChoiceZone"));
+            String definedChoiceZone = sa.getParam("ChoiceZone");
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/ability/ChooseCardAi.java:66]: try to use classes own helpers instead of manually coding like `listValueOf` here

Comment (Agetian) [reply forge-ai/src/main/java/forge/ai/ability/ChooseCardAi.java:66]: Done :) Thanks for the heads up, didn't realize listValueOf was a thing :)

Comment (tool4ever) [reply forge-ai/src/main/java/forge/ai/ability/ChooseCardAi.java:65]: Still seems way overcomplicated :p
why not just have the code look like this?
```
        List<ZoneType> choiceZone = Lists.newArrayList(ZoneType.Battlefield);
        if (sa.hasParam("ChoiceZone")) {
            choiceZone = ZoneType.listValueOf(sa.getParam("ChoiceZone"));
        }
        CardCollectionView choices = game.getCardsIn(choiceZone);
```

Comment (Agetian) [reply forge-ai/src/main/java/forge/ai/ability/ChooseCardAi.java]: Ah, indeed :) Thanks a lot for the suggestion! Implemented.

**Comment by Hanmac** [comment]
I wonder if AI shouldn't use Draw Effects (without Discard before or After),
if the resulting Draw would get them over their hand discard on their turn?

Like, check for: (pseudocode)
```
if (!handSizeUnlimited && numDraw + cardsInHand > handSize) {
  return false;
}
```

**Comment by Agetian** [comment]
Prolly a good idea, yeah, though it's a bit tricky since the AI may be able to use the said card on its turn and get some advantage from having the extra card :/ I'll need to think how to best do this. Prolly a point for the next MR :)

**Review by Hanmac — APPROVED** [review]

**Pull request merged**